### ISSUE TITLE
add another update mode: periodically find and execute update.sh scripts in nixpkgs and make pr if some file changes

### ISSUE THREAD

**Issue by srghma** [ryantm/nixpkgs-update#117]
Problem:

there are lot of `update.sh` scripts in nixpkgs https://github.com/NixOS/nixpkgs/search?q=filename%3Aupdate.sh&unscoped_q=filename%3Aupdate.sh 

for exaple for [hubstaff](https://github.com/srghma/nixpkgs/blob/cb2b80c1ba4b27673a8aab1ec78628fd9a30cbbe/pkgs/applications/misc/hubstaff/update.sh) I've made script that:
- finds latest version of hubstaff package
- fetches it using nix-prefetch-url to calculate sha256
- and writes file `revision.json` containing all required inputs for derivation

but here is the problem: each one of `update.sh` scripts should be run by hand

Proposal:
make nixpkgs-update bot:
- periodically (once in a day for example) find and execute each `update.sh` script in nixpkgs repo
- IF `update.sh` script changes some file THEN make a pr

Pros:
- `update.sh` can execute some complex logic
- can find new versions before they propagate to Repology

Cons: 
- may be computationally expensive
- `update.sh` may execute some dangerous code
- should run periodically

**Comment by jtojnar** [comment]
The issue with random update scripts in the repo is that they have vastly different expectations of how they should be called. [`passthru.updateScript`](https://nixos.org/nixpkgs/manual/#ssec-stdenv-attributes
) is slightly more predictable update method.

Also see https://github.com/ryantm/nixpkgs-update/issues/12

**Comment by jtojnar** [comment]
Another related issue: https://github.com/NixOS/nixpkgs/issues/50665#issuecomment-439776956

**Comment by srghma** [comment]
to summarize there is:
- `passthru.updateScript` that finds new version, calculates sha256 and updates parent package file (usually using https://github.com/NixOS/nixpkgs/blob/master/pkgs/common-updater/scripts/update-source-version it finds file by `meta.position` https://github.com/NixOS/nixpkgs/blob/ffd4d3301f1188fd8f9e6205d08b5a9f8daf5d81/pkgs/common-updater/scripts/update-source-version#L72) (N.B. the `meta` attributes does not propagate to `.drv` file`)
- the https://github.com/NixOS/nixpkgs/blob/master/maintainers/scripts/update.nix finds packages with `passthru.updateScript` and runs them

TODO:
- find examples of using `passthru.updateScript` outside of nixpkgs
- find examples of using `passthru.updateScript` that generates file, relative to parent package file (like in case of hubstaff - `revision.json` file OR `node-modules.nix` file using `node2nix`

**Comment by jtojnar** [comment]
> * find examples of using `passthru.updateScript` outside of nixpkgs

https://github.com/mozilla/nixpkgs-mozilla/blob/2ff87ba7e9c2cea747a88545dfe13702e7386628/pkgs/servo/default.nix#L96-L101

However, we are disabling `m/s/update.nix` for overlays in https://github.com/NixOS/nixpkgs/pull/62478

> * find examples of using `passthru.updateScript` that generates file, relative to parent package file (like in case of hubstaff - `revision.json` file OR `node-modules.nix` file using `node2nix`

https://github.com/NixOS/nixpkgs/blob/ffd4d3301f1188fd8f9e6205d08b5a9f8daf5d81/pkgs/games/dwarf-fortress/game.nix#L89-L92

**Comment by srghma** [comment]
@jtojnar tnx, especially for https://github.com/mozilla/nixpkgs-mozilla/blob/2ff87ba7e9/update.nix

I have a question - how would you add `updateScript` to https://github.com/srghma/easy-purescript-nix/blob/master/easy-dhall/default.nix so that it doesnt fetch source from github when I run `update.nix`?

**Comment by jtojnar** [comment]
What do you mean? Why would update.nix try to update that? We are disabling overlays in https://github.com/NixOS/nixpkgs/pull/62478

**Comment by srghma** [comment]
Sorry for making it unclear

I want to add https://github.com/mozilla/nixpkgs-mozilla/blob/2ff87ba7e9/update.nix to https://github.com/srghma/easy-purescript-nix because it (tentatively) would allow:
- to write update scripts with custom dependencies (for example `update.sh` could use custom node2nix package)
- use https://github.com/mozilla/nixpkgs-mozilla/blob/2ff87ba7e9/update.nix

but here is the downside

in case of https://github.com/srghma/easy-purescript-nix/blob/master/easy-dhall/default.nix if I add `updateScript` like

```

**Comment by srghma** [comment]
@jtojnar I mean the `updateScript` approach is limited for use outside nixpkgs

for example:

IF I add `updateScript` to https://github.com/srghma/easy-purescript-nix/blob/62a4e0393f0b7de0e7cff005f931169268b86f15/easy-dhall/default.nix like this 

```
{ pkgs ? import <nixpkgs> {} }:

let
  src = pkgs.fetchFromGitHub (
    builtins.fromJSON (
      builtins.readFile ./revision.json
    )
  );
in
  (import src { inherit pkgs; }).overrideAttrs (oldAttrs: rec {
    passthru.updateScript = ./update.sh;
  })
```

(as you see I use `import from derivation`  here on `import src { inherit pkgs; }`)

AND if I run `update.nix` THEN it will fetch source from github (`src = pkgs.fetchFromGitHub`) EVEN THOUGHT this is redundant (because I'm trying to update ./revision.json, `./update.sh` should update `./revision.json`)

**Comment by ryantm** [comment]
As jtojnar said, the scripts we run need to obey a contract to be viable. passthru.updateScript can hopefully be that. (Though, I've noticed there are inconsistencies about whether the script will create a git commit or not.) I'm closing this in favor of #182.

**Comment by jtojnar** [comment]
It should never create commits since it can run in parallel. (We will need https://github.com/NixOS/nixpkgs/pull/59372 for that.) If you find such package, please report it as an issue.
